### PR TITLE
Url parameters input validation

### PR DESCRIPTION
**Pull request by sashaCher**
### What does this PR do?

Introduces a layer of input validation for uri based parameters

#### Change log
- [ ] The CHANGELOG has been updated, or
- [X] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [X] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [X] This PR does not require updating any documentation

#### API Changes
- [ ] The [OpenAPI spec](https://github.com/cyberark/conjur-openapi-spec) has been updated to meet new API changes (or an issue has been opened), or
- [X] The changes in this PR do not affect the Conjur API

### CODE REVIEW

File: app/controllers/authenticate_controller.rb [hunk 1]
```diff
@@ -122,6 +123,21 @@ def authenticator_input
     )
   end
 
+  def tiny_authenticator_input
```
Comment (orenbm) [inline app/controllers/authenticate_controller.rb:126]: so id this the same as `authenticator_input` but without credentials? can we call it `authenticator_input_without_credentials` to be more explicit?

File: app/domain/authentication/authn_jwt/authenticator.rb [hunk 2]
```diff
@@ -25,6 +25,8 @@ def call
       def create_authentication_parameters
         @logger.debug(LogMessages::Authentication::AuthnJwt::CREATING_AUTHENTICATION_PARAMETERS_OBJECT.new)
         @authentication_parameters = Authentication::AuthnJwt::AuthenticationParameters.new(@authenticator_input)
+        @logger.debug(LogMessages::Authentication::AuthnJwt::FetchingTokenString.new)
+        @authentication_parameters.jwt_token = @jwt_configuration.create_authentication_parameters(@authentication_parameters)
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/authenticator.rb:29]: i would move the `jwt_token` to the constructor instead of using `attr_accessor`.

Comment (orenbm) [reply app/domain/authentication/authn_jwt/authenticator.rb:29]: oh damn i now see that the `authentication_parameters` contains a `jwt_token` to contains the `authentication_parameters`. is that correct?

File: app/domain/authentication/authn_jwt/configuration_jwt_generic_vendor.rb [hunk 3]
```diff
@@ -2,7 +2,9 @@ module Authentication
   module AuthnJwt
     # Mock JWTConfiguration class to use it to develop other part in the jwt authenticator
     class ConfigurationJWTGenericVendor < ConfigurationInterface
+
       def initialize
+        @token_fetcher = Authentication::AuthnJwt::FetchTokenFromCredentials.new
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/configuration_jwt_generic_vendor.rb:7]: please call this param in the same name as the class it represents (`fetch_token_from_credentials`)

File: app/domain/authentication/authn_jwt/authenticator.rb (hunk 2 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/authenticator.rb]: The change is related to previous PR.

File: app/domain/authentication/authn_jwt/configuration_jwt_generic_vendor.rb (hunk 3 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/configuration_jwt_generic_vendor.rb]: The change is related to previous PR.

File: app/controllers/authenticate_controller.rb (hunk 1 above)
Comment (sashaCher) [reply app/controllers/authenticate_controller.rb]: The change is subject for next PR. I'll take your remark forward :-)

File: app/domain/authentication/authn_jwt/orchestrate_jwt_authentication.rb [hunk 4]
```diff
@@ -6,14 +6,16 @@ module AuthnJwt
 
     OrchestrateAuthentication ||= CommandClass.new(
       dependencies: {
+        uri_based_parameters_validator: Authentication::AuthnJwt::ValidateUriBasedParameters.new,
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/orchestrate_jwt_authentication.rb:9]: please call this `validate_uri_based_parameters` to match the class name

File: app/domain/authentication/authn_jwt/validate_uri_based_parameters.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
+module Authentication
+  module AuthnJwt
+
+    ValidateUriBasedParameters ||= CommandClass.new(
+      dependencies: {
+        # ValidateWebserviceIsWhitelisted calls ValidateAccountExists
+        # we call there ValidateAccountExists for better readability an safety
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/validate_uri_based_parameters.rb:7]:
```suggestion
        # we call ValidateAccountExists for better readability and safety
```

File: app/domain/authentication/authn_jwt/validate_uri_based_parameters.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        validate_account_exists: ::Authentication::Security::ValidateAccountExists.new,
+        validate_webservice_is_whitelisted: Security::ValidateWebserviceIsWhitelisted.new
+      },
+      inputs: %i[authenticator_input enabled_authenticators]
+    ) do
+
+      # authenticator_input attributes
+      #  :account - from uri - has explicit validation
+      #  :authenticator_name - from uri - has implicit validation with service_id
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/validate_uri_based_parameters.rb:16]: what do you mean here? where is this validated?

File: app/domain/authentication/authn_jwt/validate_uri_based_parameters.rb [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+      #  :username - can be from uri
+      #    - validation is specific for authenticator
+      #    - in jwt case undefined until token signature validation
+      #  :credentials - the web request body - validation is specific for authenticator
+      #  :client_ip - from metadata or request - depends username
+      #  :request - the web request object
+
+      def call
+        validate_account_exists
+        validate_webservice_is_whitelisted
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/validate_uri_based_parameters.rb:27]: i think that this validation is currently in the JWT authenticator class. shouldn't this PR remove it?

File: app/domain/authentication/authn_jwt/validate_uri_based_parameters.rb (hunk 6 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/validate_uri_based_parameters.rb:16]: According my understanding when `webservice` resource is created it automatically "bounded" to DAL means if `authenticator_name` and `service_id` does not exist in DB resource creation will failed.
@orenbm Please fix me if I'm wrong.

File: app/domain/authentication/authn_jwt/validate_uri_based_parameters.rb (hunk 7 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/validate_uri_based_parameters.rb:27]: No it does not exist in JWT authenticator class

File: app/domain/authentication/authn_jwt/orchestrate_jwt_authentication.rb (hunk 4 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/orchestrate_jwt_authentication.rb:9]: Done

File: app/domain/authentication/authn_jwt/validate_uri_based_parameters.rb (hunk 5 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/validate_uri_based_parameters.rb:7]: Done

File: app/domain/authentication/authn_jwt/validate_uri_based_parameters.rb (hunk 6 above)
Comment (orenbm) [reply app/domain/authentication/authn_jwt/validate_uri_based_parameters.rb:16]: i don't know off by heart, please verify. also - in such a case, what error is written to the log? is it clear?

**Review by orenbm — APPROVED** [review]

**Pull request merged**